Initial Plugin

kokrui opened the pull request:
As per https://github.com/trailofbits/MUI/issues/32 creates an initial MUI-Ghidra plugin.

Currently still a draft PR, pending accompanying documentation. Adds 2 components to Ghidra, a setup component and a log.

![image](https://user-images.githubusercontent.com/29654756/146400647-2bf2d4fa-8991-4835-8b55-7f3c8d04557d.png)

kokrui committed:
init GhidraDev plugin

kokrui committed:
manticore options UI elements

kokrui committed:
basic output UI, run manticore

kokrui committed:
create log component

kokrui committed:
remove outputpanel

kokrui committed:
run manticore in separate thread, log stdout

kokrui committed:
clean up UI, prefill manticore args

kokrui committed:
format

kokrui committed:
update desc

kokrui committed:
refactor commandArgs -> manticoreArgs

kokrui committed:
clean debug statements

kokrui committed:
clear log on clear button click

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/trailofbits/MUI-Ghidra?pullRequest=1) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/trailofbits/MUI-Ghidra?pullRequest=1) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/trailofbits/MUI-Ghidra?pullRequest=1) it.</sub>

kokrui committed:
remove currently unused stubs

kokrui committed:
disable runBtn while manticore running

kokrui committed:
Init Docs for Initial Plugin

kokrui committed:
include native manticore binary for linux

kokrui committed:
Merge branch 'initial-plugin' of https://github.com/kokrui/MUI-Ghidra into initial-plugin

kokrui commented:
Tasks 1 through 6 on https://github.com/trailofbits/MUI/issues/32 should now be covered by this PR - I don't however have a way to create a Mac-compatible executable for Task 7 at the moment.

ekilmer commented on MUI/.project:
```diff
@@ -0,0 +1,30 @@
+		<link>
+			<name>Ghidra</name>
+			<type>2</type>
+			<location>/home/kr-kal/Desktop/ghidra_10.1_PUBLIC</location>
```
I think this file might need to be added to `.gitignore` because this path is user-dependent

ekilmer commented on MUI/.pydevproject:
```diff
@@ -0,0 +1,137 @@
+            
+    
+    <pydev_pathproperty name="org.python.pydev.PROJECT_EXTERNAL_SOURCE_PATH">
+        <path>/home/kr-kal/Desktop/ghidra_10.1_PUBLIC/Ghidra/Processors/68000/lib/68000.jar</path>
```
I think this file might need to be added to `.gitignore` because this path is user-dependent

ekilmer commented on MUI/src/main/java/mui/MUIPlugin.java:
```diff
@@ -0,0 +1,56 @@
+@PluginInfo(
+	status = PluginStatus.UNSTABLE,
+	packageName = "MUI",
+	category = PluginCategoryNames.EXAMPLES,
```
Is there a better enum for MUI? Doesn't matter too much at this point, but might as well change it so we don't forget

kokrui replied:
You can see the available categories [here](https://github.com/NationalSecurityAgency/ghidra/blob/master/Ghidra/Framework/Project/src/main/java/ghidra/app/plugin/PluginCategoryNames.java)

Would `ANALYSIS` be most appropriate?

ekilmer replied:
Yeah, I think `ANALYSIS` would fit pretty well.

ekilmer committed:
Add gradle install tasks

The following Makefile-like commands should be usable

install-mui:
 gradle install

reinstall-mui:
 gradle -PMUI_AUTO_REMOVE install

uninstall-mui:
 gradle -PMUI_AUTO_REMOVE uninstallPreviousMUI

ekilmer reviewed (changes requested):
Good start!

I noticed a few weird things that we should try fixing while the codebase is still small.

There seems to be some weird overlapping of text here. Is there a way to either set a minimum window size or potentially just move the "MUI Setup" text into the blue border text?
<img width="118" alt="Screen Shot 2021-12-27 at 10 10 02 AM" src="https://user-images.githubusercontent.com/816362/147484475-a18c4e4e-f5e1-488f-a83d-d583f42ba47c.png">

A small nit, but would be nice to fix is that there is a border running through the tab buttons here
<img width="194" alt="Screen Shot 2021-12-27 at 10 10 29 AM" src="https://user-images.githubusercontent.com/816362/147484619-3cb737f2-5e2e-4a82-bd5a-c564f9ef4299.png">
 
I also don't think we actually need a Mac binary right now, but when I tried clicking the "Run" button without any sort of `manticore` command on path, nothing really happened and the "MUI Log" just said `Manticore execution complete.`. There should be an error message for the case where the command or binary is not found.

Also, I think if we want to keep the logs of every Manticore run in the MUI Log page, we should indicate the start and end of the logs for each run with maybe a counter or something. Otherwise, we should clear the log output after each press of the "Run" button. Potentially, a new "MUI Log" tab is created for each run, for extra points 😄.

kokrui committed:
update PluginInfo category

kokrui committed:
gitignore user-dependent files, add dev instructions

kokrui committed:
clean up MUI Setup widget title

kokrui committed:
remove some unused code

kokrui committed:
actually remove user-dependent files

kokrui committed:
refactor to MUISetupProvider

kokrui committed:
add tabs in log, allow multiple manticore threads, modularise code, log errors

kokrui commented:
now supports concurrent manticore processes (and a separate tab for each process). small demo:
![mui-tabs](https://user-images.githubusercontent.com/29654756/148135011-9a32889c-a0f5-44f3-bef4-9c8259a308db.gif)

currently, tab titles are just the timestamp where manticore was run - couldn't think of something more natural/appropriate, do you guys have a preference?

additionally,
- the smaller issues above have been fixed (UI bugs, appropriate enum, gitignore)
- manticore stderr will now be logged, as well as any errors encountered by the plugin itself 
- the code has been somewhat modularised to make the upcoming PR a bit easier to handle, though i'm expecting the structure to change to fit in additional features and to eventually somewhat converge towards the binja plugin/the cross-platform interface
- the plugin, for now, no longer attempts to use manticore on PATH, and only runs manticore if there is a bundled manticore binary. As discussed, a feature to choose which manticore to use (bundled/PATH/specify dir) will probably be implemented in the future. 
- I've been able to use the bundled manticore binary on a fresh ubuntu and debian VM, couldn't figure out why it would have failed on your fresh Linux box @ekilmer

ekilmer commented:
Regarding the `manticore` binary that is checked in. I think I figured it out. If you take a look at the first line of the `manticore` binary, it has the following line:
```
head -n 1 ./MUI/os/linux_x86_64/manticore
#!/usr/bin/env python3
```
On my Fedora VM, that points to Python 3.10, so I get the following error
```
$ ./MUI/os/linux_x86_64/manticore --version
Traceback (most recent call last):
  File "/home/ekilmer/src/MUI-Ghidra/./MUI/os/linux_x86_64/manticore/_bootstrap/__init__.py", line 76, in import_string
  File "/home/ekilmer/.shiv/manticore_790abfe3c4cf98e92421049a09dd3ac958109a4a5adba4ba00c16129c5253e99/site-packages/manticore/__init__.py", line 10, in <module>
    from .ethereum.manticore import ManticoreEVM
  File "/home/ekilmer/.shiv/manticore_790abfe3c4cf98e92421049a09dd3ac958109a4a5adba4ba00c16129c5253e99/site-packages/manticore/ethereum/__init__.py", line 2, in <module>
    from .abi import ABI
  File "/home/ekilmer/.shiv/manticore_790abfe3c4cf98e92421049a09dd3ac958109a4a5adba4ba00c16129c5253e99/site-packages/manticore/ethereum/abi.py", line 6, in <module>
    import sha3
  File "/home/ekilmer/.shiv/manticore_790abfe3c4cf98e92421049a09dd3ac958109a4a5adba4ba00c16129c5253e99/site-packages/sha3.py", line 8, in <module>
    from _pysha3 import keccak_224, keccak_256, keccak_384, keccak_512
ModuleNotFoundError: No module named '_pysha3'
```
and when I search through the `manticore` binary again, I see references to Python 3.9 on line 17
```
$ sed -n '17p' MUI/os/linux_x86_64/manticore | strings -s ' '
}exU site-packages/_distutils_hack/override.py ,.)- /K-* &@m& site-packages/_pysha3.cpython-39-x86_64-linux-gnu.so d&Qb
```

All of this is to say that I should use Python 3.9 to fix the errors I'm seeing and take advantage of the pre-built binary. However, I think a better solution would be to use shiv's `--python $(which python3.9)` during creation, which should set the first line of the `manticore` binary as `#!/usr/bin/env python3.9` 🤞

kokrui committed:
bundled manticore binary forces py3.9

kokrui committed:
update readme, require python3.9

ekilmer committed:
Add code formatting and linting

ekilmer committed:
Rebuild Manticore binary with native dependencies
